### ISSUE TITLE
xdg module should update `home.sessionPath`

### ISSUE THREAD

**Issue by sellout** [nix-community/home-manager#3357]
The [XDG spec](https://specifications.freedesktop.org/basedir-spec/latest/ar01s03.html) says “User-specific executable files may be stored in $HOME/.local/bin. Distributions should ensure this directory shows up in the UNIX $PATH environment variable, at an appropriate place.”

So, xdg.nix should check `home.sessionPath` for `$HOME/.local/bin` (including paths that have the same expansion), and if it’s not there, prepend it to the list.

This could be controlled by a new attribute, `xdg.sessionPath.enable`, which (IMO) should default to `true`.

**Comment by sellout** [comment]
Even though there isn’t an `XDG_BIN_HOME` variable, it would also be nice if there was an `xdg.binFile` attribute for defining files to be linked into `$HOME/.local/bin`.

I think the lack of environment variable is because applications shouldn’t explicitly use that path to find binaries, but instead rely on `PATH` being set correctly. However, home-manager’s use case is different, and using that path is in the spirit of the spec.

**Comment by tapeinosyne** [comment]
I agree that a neat way to symlink user executables is desirable, but feel that the XDG specification is a bit undercooked here: it would be awkard to have a single XDG location that, unlike the others, cannot configured. For example, a user who eschewed `.local` and placed both `XDG_DATA_HOME` and `STATE_HOME` elsewhere would end up with an undesired `.local/bin` in their home.

I suppose Home Manager could take the liberty to introduce an `xdg.binHome` variable defaulting to `$HOME/.local/bin`, but at that point it might be more sensible to sidestep the specification and create an independent module or option.

If the maintainers agree on a solution, I would be happy to contribute the feature (although that would likely require some guidance, since I have no experience with Nix beyond the trivial).

**Comment by ncfavier** [comment]
We could just mirror what NixOS does with [environment.localBinInPath](https://github.com/NixOS/nixpkgs/blob/nixos-unstable/nixos/modules/config/shells-environment.nix#L129-L135). Default to false, because we don't like state (in that sense we don't fall into the typical "distributions" the XDG spec talks about).

**Comment by sellout** [comment]
I’ll put up a PR for this, adding `xdg.sessionPath.enable`. I’d like it to default to `true`, but @ncfavier, I’m curious about why you want it to default to `false` – do you mean specifically `home.stateVersion` (if it defaults to `true`, I suppose it does have to be conditional on that), or more generally does having `.local/bin` create/encourage state?

I’m thinking
```nix
{
  xdg.localBin = {
    sessionPath = true; # add to $PATH
    source = ./home/bin; # like in `file-type`, symlink to here
    recursive = true; # like in `file-type`, to allow manually adding binaries to the directory
    target = somewhere; # like in `file-type`, defaults to $HOME/.local/bin
  };
}
```
would be great for my usage.

Really, other than `text`, it could use all of file-type … wonder if file-type could be modularized a bit, have a directory-type (without `text`) a file-type (without `recursive`), and type that can be either (to be used in the current locations).

**Comment by ncfavier** [comment]
> more generally does having `.local/bin` create/encourage state

That's what I meant.
